File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while(fast != null)
        {
            fast = fast.next;
            if(fast == null)   // case of odd nodes
                return slow;
            slow = slow.next;
            fast = fast.next;
        }
        return slow;    // case of even nodes
    }
}

File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* med = head;
        ListNode* tail = head;
        while(tail!=nullptr && tail->next != nullptr) {
            tail = tail->next->next;
            med = med->next;
        }
        return med;
    }
};